modules/mini-games/speedrun.lua

```diff
@@ -0,0 +1,537 @@
+    end
+
+    Task.queue_task(surface_generator, remaining, team_count)
+end
+
+--- Called once enough participants are present to start the game and map generation is done
+local function start()
+    -- Chart the start area for all teams
+    local removed = {}
+    for name, force in pairs(forces) do
+        if #force.players > 0 then
+            force.share_chart = true
+            force.chart(surfaces[name], {{x = 80, y = 80}, {x = -80, y = -80}})
+        else
+            game.merge_forces(force, game.forces.player)
+            game.delete_surface(surfaces[name])
+            removed[name] = true
+        end
+    end
+
+    -- Remove references to invalid forces
+    for name in pairs(removed) do
+        progress[name] = nil
+        surfaces[name] = nil
+        forces[name] = nil
+    end
+
+    -- Added all the teams to the progress table
+    local tooltip = goals[primitives.target].goal
+    for _, player in pairs(game.players) do
+        Gui.toggle_left_element(player, timer_container, true)
+        local container = Gui.get_left_element(player, timer_container)
+        local progress_table = container.progress_table
+        container.timer.tooltip = tooltip
+        for name in pairs(forces) do team_entry(progress_table, name) end
+    end
+end
+
+----- Game Stop and Close -----
+
+--- Called to stop the game and return the results to be saved
+local function stop()
+    local scores, ctn = {}, 0
+    -- Get all the data needed to write results
+    for name, team in pairs(progress) do
+        ctn = ctn + 1
+        local names = {}
+        for index, player in ipairs(forces[name].players) do names[index] = player.name end
+        scores[ctn] = { name, math.floor(team[1]/team[2]*1000)/10, names }
+    end
+
+    -- Sort by team progress
+    table.sort(scores, function(a, b)
+        return a[2] > b[2]
+    end)
+
+    -- Format the results table
+    local results, names = {}, {}
+    for _, team in ipairs(scores) do
+        local score = team[2]
+        local last = #results
+        local up_result = results[last]
+        if up_result and up_result.score == score then
+            names[last] = names[last]..', '..team[1]
+            local players = up_result.players
+            local offset = #players
+            for index, player in ipairs(team[3]) do
+                players[offset+index] = player
+            end
+        else
+            names[last+1] = team[1]
+            results[last+1] = { place = last+1, score = score, players = team[3] }
+        end
+    end
+
+    Mini_games.print_results(results, 'percent', names)
+    return results
+end
+
+--- Final function called by the mini game core in order to clean up
+local function close()
+    -- Remove the surfaces
+    for _, surface in pairs(surfaces) do
+        game.delete_surface(surface)
+    end
+
+    -- Remove the forces
+    for _, force in pairs(forces) do
+        game.merge_forces(force, game.forces.player)
```

Comment:
is their a reason to remove all forces and not just keep them for the next game?

Reply:
It is easier than removing all the entities on a surface and resetting the technology and benifits for each force.